rustworkx-core/src/collect_bicolor_runs.rs

```diff
@@ -0,0 +1,203 @@
+use std::fmt::{Debug, Display, Formatter};
+
+use petgraph::algo;
+use petgraph::visit::Data;
+use petgraph::data::DataMap;
+use petgraph::visit::{EdgeRef, GraphBase, IntoNeighborsDirected, IntoNodeIdentifiers, NodeIndexable, Visitable, IntoEdgesDirected};
+
+
+// Taken from Kevin's PR, but we probably don't need the enum (no MergeError either)
+// TODO: clean up once the code compiles
+#[derive(Debug)]
+pub enum CollectBicolorError {
+    DAGWouldCycle,
+}
+
+impl Display for CollectBicolorError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CollectBicolorError::DAGWouldCycle => fmt_dag_would_cycle(f),
+        }
+    }
+}
+
+impl Error for CollectBicolorError {}
+
+#[derive(Debug)]
+pub enum CollectBicolorSimpleError<E: Error> {
+    DAGWouldCycle,
+    MergeError(E), //placeholder, may remove if not used
+}
+
+impl<E: Error> Display for CollectBicolorSimpleError<E> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CollectBicolorSimpleError::DAGWouldCycle => fmt_dag_would_cycle(f),
+            CollectBicolorSimpleError::MergeError(ref e) => fmt_merge_error(f, e),
+        }
+    }
+}
+
+impl<E: Error> Error for CollectBicolorSimpleError<E> {}
+
+fn fmt_dag_would_cycle(f: &mut Formatter<'_>) -> std::fmt::Result {
+    write!(f, "The operation would introduce a cycle.")
+}
+
+fn fmt_merge_error<E: Error>(f: &mut Formatter<'_>, inner: &E) -> std::fmt::Result {
+    write!(f, "The prov failed with: {:?}", inner)
+}
+/// Collect runs that match a filter function given edge colors
+///
+/// A bicolor run is a list of group of nodes connected by edges of exactly
+/// two colors. In addition, all nodes in the group must match the given
+/// condition. Each node in the graph can appear in only a single group
+/// in the bicolor run.
+///
+/// :param PyDiGraph graph: The graph to find runs in
+/// :param filter_fn: The filter function to use for matching nodes. It takes
+///     in one argument, the node data payload/weight object, and will return a
+///     boolean whether the node matches the conditions or not.
+///     If it returns ``True``, it will continue the bicolor chain.
+///     If it returns ``False``, it will stop the bicolor chain.
+///     If it returns ``None`` it will skip that node.
+/// :param color_fn: The function that gives the color of the edge. It takes
+///     in one argument, the edge data payload/weight object, and will
+///     return a non-negative integer, the edge color. If the color is None,
+///     the edge is ignored.
+///
+/// :returns: a list of groups with exactly two edge colors, where each group
+///     is a list of node data payload/weight for the nodes in the bicolor run
+/// :rtype: list
+pub fn collect_bicolor_runs<G, F, C, B, E>(
+    graph: G,
+    filter_fn: F,
+    color_fn: C,
+) -> Result<Vec<Vec<&<G as Data>::NodeWeight>>, CollectBicolorSimpleError<E>> //OG type: PyResult<Vec<Vec<PyObject>>>
```

Comment:
Instead of returning `NodeWeight` here I think it would be simpler to just return something like:

```
impl Iterator<Item=Vec<G::NodeId>>
```
(or something like that I might have made a mistake in the exact syntax) Or alternatively a `Vec<Vec<G::NodeId>>`

The idea is to return an iterator of node indices for each run. This isn't exactly what the python api was returning, but retunring the node ids will let us easily iterate and map that to the weights for python, but for other rust consumer (especially Qiskit in the future) this will be lighter weight to work with because it's a simple vec get to go from an index to a weight. 

Then using an iterator would be a more rust native interface to return an iterator of the runs